plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,673 @@
+	/* first two bytes are report id */
+	report.report_id = GINT16_TO_LE(FU_SYNAPTICS_CAPE_DEVICE_GOLEM_REPORT_ID);
+
+	if (!fu_memcpy_safe((guint8 *)&report.cmd,
+			    sizeof(report.cmd),
+			    0, /* dst */
+			    (const guint8 *)req,
+			    sizeof(*req),
+			    0, /* src */
+			    sizeof(*req),
+			    error))
+		return FALSE;
+
+	/* sets data length to MAX for any GET commands */
+	if (FU_SYNAPTICS_CMD_GET_FLAG & report.cmd.cmd_id) {
+		is_get = TRUE;
+		report.cmd.data_len = GINT16_TO_LE(FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN);
+	} else {
+		report.cmd.data_len = GINT16_TO_LE(report.cmd.data_len);
+	}
+
+	report.cmd.cmd_id = GUINT32_TO_LE(report.cmd.cmd_id);
+	report.cmd.module_id = GUINT32_TO_LE(report.cmd.module_id);
+
+	if (!fu_synaptics_cape_device_set_report(self, &report, error)) {
+		g_prefix_error(error, "failed to send: ");
+		return FALSE;
+	}
+	if (delay_us > 0)
+		g_usleep(delay_us);
+
+	/* wait for the command to complete */
+	for (; elapsed_ms < FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_TIMEOUT;
+	     elapsed_ms += FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL) {
+		if (!fu_synaptics_cape_device_get_report(self, &report, error))
+			return FALSE;
+		if (report.cmd.reply)
+			break;
+		g_usleep(FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL * 1000);
+	}
+
+	if (!report.cmd.reply) {
+		g_set_error(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "send command time out");
+		return FALSE;
+	}
+
+	/* copy returned data if it is GET command */
+	if (is_get) {
+		req->data_len =
+		    (gint16)fu_common_read_uint16((guint8 *)&report.cmd, G_LITTLE_ENDIAN);
+
+		for (int i = 0; i < FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN; i++)
+			req->data[i] = GUINT32_FROM_LE(report.cmd.data[i]);
+	}
+
+	return fu_synaptics_cape_device_rc_set_error(&report.cmd, error);
+}
+
+/* a simple version of sendcmd_ex without returned data */
+static gboolean
+fu_synaptics_cape_device_sendcmd(FuSynapticsCapeDevice *self,
+				 const guint32 module_id,
+				 const guint32 cmd_id,
+				 const guint32 *data,
+				 const guint32 data_len,
+				 const gulong delay_us,
+				 GError **error)
+{
+	FuCapCmd cmd = {0};
+	const guint32 dataszbyte = data_len * FU_SYNAPTICS_CAPE_WORD_IN_BYTES;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	cmd.cmd_id = cmd_id;
+	cmd.module_id = module_id;
+
+	if (data_len != 0 && data != NULL) {
+		cmd.data_len = data_len;
+		if (!fu_memcpy_safe((guint8 *)cmd.data,
+				    sizeof(cmd.data),
+				    0, /* dst */
+				    (const guint8 *)data,
+				    dataszbyte,
+				    0, /* src */
+				    dataszbyte,
+				    error))
+			return FALSE;
+	}
+	return fu_synaptics_cape_device_sendcmd_ex(self, &cmd, delay_us, error);
+}
+
+static void
+fu_synaptics_cape_device_to_string(FuDevice *device, guint idt, GString *str)
+{
+	FuSynapticsCapeDevice *self = FU_SYNAPTICS_CAPE_DEVICE(device);
+
+	g_return_if_fail(FU_IS_SYNAPTICS_CAPE_DEVICE(self));
+
+	fu_common_string_append_ku(str, idt, "active_partition", self->active_partition);
```

Comment:
`fu_common_string_append_ku(str, idt, "ActivePartition", self->active_partition);` please

Reply:
oops, I don't know why this is back. I am working on it.

Reply:
That might have been one of my over-zealous search/replaces. If so, apologies.